src/engine/sync/midimaster.cpp

```diff
@@ -0,0 +1,131 @@
+#include <QtDebug>
+
+#include "engine/sync/midimaster.h"
+
+#include "engine/sync/enginesync.h"
+#include "controllers/midi/midisourceclock.h"
+#include "controllinpotmeter.h"
+#include "controlobject.h"
+#include "controlpushbutton.h"
+#include "configobject.h"
+#include "util/time.h"
+
+MidiMasterClock::MidiMasterClock(const char* pGroup, SyncableListener* pEngineSync)
+        : m_group(pGroup),
+          m_pEngineSync(pEngineSync),
+          m_mode(SYNC_NONE) {
+    m_pMidiSourceClockBpm.reset(new ControlObject(ConfigKey(pGroup, "bpm")));
+    m_pMidiSourceClockLastBeatTime.reset(
+            new ControlObject(ConfigKey(pGroup, "last_beat_time")));
+    m_pMidiSourceClockBeatDistance.reset(
+            new ControlObject(ConfigKey(pGroup, "beat_distance")));
+
+    m_pMidiSourceClockRunning.reset(
+            new ControlPushButton(ConfigKey(pGroup, "run")));
+
+    m_pSyncMasterEnabled.reset(
+            new ControlPushButton(ConfigKey(pGroup, "sync_master")));
+    m_pSyncMasterEnabled->setButtonMode(ControlPushButton::TOGGLE);
+    m_pSyncMasterEnabled->connectValueChangeRequest(
+            this, SLOT(slotSyncMasterEnabledChangeRequest(double)),
+            Qt::DirectConnection);
+
+    m_pMidiSourceClockSyncAdjust.reset(
+            new ControlLinPotmeter(ConfigKey(pGroup, "sync_adjust"), -.5, .5,
+                                   0.1, 0.01, /*allow oob*/ true));
+}
+
+MidiMasterClock::~MidiMasterClock() {
+};
+
+void MidiMasterClock::notifySyncModeChanged(SyncMode mode) {
+    // Syncable has absolutely no say in the matter. This is what EngineSync
+    // requires. Bypass confirmation by using setAndConfirm.
+    m_mode = mode;
+    m_pSyncMasterEnabled->setAndConfirm(mode == SYNC_MASTER);
+}
+
+void MidiMasterClock::notifyOnlyPlayingSyncable() {
+    // No action necessary.
+}
+
+void MidiMasterClock::requestSyncPhase() {
+    // TODO: maybe tell MidiSourceClock to reset which tick is the beat tick??
+    // but really, it's a read-only clock.
+}
+
+void MidiMasterClock::slotSyncMasterEnabledChangeRequest(double state) {
+    bool currentlyMaster = getSyncMode() == SYNC_MASTER;
+
+    if (state > 0.0) {
+        if (currentlyMaster) {
+            // Already master.
+            return;
+        }
+        m_pEngineSync->requestSyncMode(this, SYNC_MASTER);
+    } else {
+        // TODO: midi follower (clock out?)
+        m_pEngineSync->requestSyncMode(this, SYNC_NONE);
+    }
+}
+
+double MidiMasterClock::getBeatDistance() const {
+    qint64 last_beat = static_cast<qint64>(m_pMidiSourceClockLastBeatTime->get());
+    double raw_percent = MidiSourceClock::beatFraction(last_beat, Time::elapsed(),
+                                                       m_pMidiSourceClockBpm->get());
+    raw_percent += m_pMidiSourceClockSyncAdjust->get();
```

Comment:
To be more specific, although the adjustment is one-time, while the user is adjusting it the calculated values will be incorrect unless the entire ringbuffer is munged at the same time, which seems tricky and bug-prone.  Also, it seems very hacky to munge the result of ->now() for all callers throughout mixxx just to handle the sync of one particular midi controller.  For instance, there may be more than one midi controller generating timecode data and they may need different offsets.

Btw you didn't account for possible midi offset / drift in the chain upstream from the midi receive queue, so that's another source of possible delay.